File: aoc2220.cc
#include "iostream"
#include "vector"
#include "deque"
#include "algorithm"
#include "cstdlib" // abs
#include "cassert"

#define ll long long

using	namespace std;

// part 1 :: 5842 @ 613 . 5376 @ 611 > hi

ll	solve(vector<ll>, vector<ll>, ll, int);

int	main()
{
	ll n, i;
	vector<ll> a;
	while (!cin.eof() && cin >> n)
		a.push_back(n);
	n = a.size();
	vector<ll> idx;
	i = -1;
	while (++i < n)
		idx.push_back(i);
	ll key = 811589153;
	ll res = solve(a, idx, 1, 1);
	ll res2 = solve(a, idx, key, 10);

	assert(res == 3 || res == 1087);
	assert(res2 == 1623178306 || res2 == 13084440324666);

	cout << "Star 1: " << res << endl;
	cout << "Star 2: " << res2 << endl;
}

ll	solve(vector<ll> a, vector<ll> idx, ll key, int times)
{
	for (long long& n: a)
		n *= key;
	ll index, value, move2 = 0;
	ll res = 0;
	ll t = -1, i;
	vector<ll>::iterator it;
	while (++t < times)
	{
		i = -1;
		while (++i < (int) a.size())
		{
			if ((it = find(idx.begin(), idx.end(), i)) == idx.end()) // assert !end
				return 0 ;
			index = it - idx.begin();
			value = a[index];
			a.erase(a.begin() + index);
			ll N = index + value;
			ll S = (long long) a.size();
			if (N < 0)
				N = (((S + N) % S) + S) % S;
			else
				N = N % (a.size());
			// N = abs(((int) a.size() + N) % (int) a.size()); // old way is buggy
			//	because in c++ modulo between 2 arguments of different
			//	signs gives language-specific results.
			//	a common fix it is :
			//		(b + (a % b)) % b
			move2 = N;
			if (move2 == 0 && value)
				move2 = (long long) a.size();
			else if (move2 == (long long) a.size())
				move2 = 0;
			a.insert(a.begin() + move2, value);
			idx.erase(idx.begin() + index);
			idx.insert(idx.begin()+ move2, i);
		}
		i = find(a.begin(), a.end(), 0) - a.begin();
		ll S = (long long) a.size();
		res = a[(1000 + i) % S] + a[(2000 + i) % S] + a[(3000 + i) % S];
	}
	return res;
}

File: aoc2220.py
from typing import List
import copy

def main():
    file = open('2220.1')    
    res, res2 = day22(file)
    print('test')
    print(f'Star 1: {res} \nStar 2: {res2}\n')

    file = open('2220.0')    
    res, res2 = day22(file)
    print('data')
    print(f'Star 1: {res} \nStar 2: {res2}')

def day22(file) -> (int, int):
    line = file.read().strip().split()
    a = [int(_) for _ in line]
    idx = [i for i in range(0, len(a))]
    aa = copy.deepcopy(a)
    idxx = copy.deepcopy(idx)
    
    yield solve(a, idx, 1, 1)
    
    p2 = 811589153
    yield solve(aa, idxx, p2, 10)

def solve(a: List[int], idx: List[int], key: int, times: int) -> int:
    for i in range(len(a)):
        a[i] *= key
    res = 0
    for t in range(times):
        for i in range(len(a)):
            if i not in idx:
                return 0
            index = idx.index(i)
            value = a.pop(index)
            N = index + value
            S = len(a)
            move2 = N % S
            if move2 == 0 and value != 0:
                move2 = len(a)
            elif move2 == len(a):
                move2 = 0
            a.insert(move2, value)
            idx.pop(index)
            idx.insert(move2, i)
        i = a.index(0)
        S = len(a)
        th = 1000
        res = a[(th + i) % S] \
            + a[(th*2+i) % S] \
            + a[(th*3+i) % S]
    return res

if __name__ == '__main__':
    main()
